File: server/loans/core/src/main/java/ee/rsx/kata/bank/loans/usecases/SsnValidation.java
package ee.rsx.kata.bank.loans.usecases;

import ee.rsx.kata.bank.loans.domain.ssn.SocialSecurityNumber;
import ee.rsx.kata.bank.loans.validation.ssn.SsnValidationResultDTO;
import ee.rsx.kata.bank.loans.validation.ssn.ValidateSocialSecurityNumber;
import jakarta.inject.Named;

import java.time.format.DateTimeParseException;

import static ee.rsx.kata.bank.loans.validation.ssn.SsnValidationResultDTO.*;

@Named
class SsnValidation implements ValidateSocialSecurityNumber {

  @Override
  public SsnValidationResultDTO on(String ssn) {
    try {
      var validSsn = new SocialSecurityNumber(ssn);
      return okResultWith(validSsn.value());
    } catch (IllegalArgumentException | IllegalStateException | DateTimeParseException e) {
      return invalidResultWith(ssn);
    }
  }
}

File: server/loans/core/src/main/kotlin/ee/rsx/kata/bank/loans/usecases/SsnValidation.kt
package ee.rsx.kata.bank.loans.usecases

import ee.rsx.kata.bank.loans.domain.ssn.SocialSecurityNumber
import ee.rsx.kata.bank.loans.validation.ssn.SsnValidationResultDTO
import ee.rsx.kata.bank.loans.validation.ssn.SsnValidationResultDTO.invalidResultWith
import ee.rsx.kata.bank.loans.validation.ssn.SsnValidationResultDTO.okResultWith
import ee.rsx.kata.bank.loans.validation.ssn.ValidateSocialSecurityNumber
import jakarta.inject.Named
import java.time.format.DateTimeParseException

@Named
internal class SsnValidation : ValidateSocialSecurityNumber {

  override fun on(ssn: String): SsnValidationResultDTO =
    try {
      val validSsn = SocialSecurityNumber(ssn)
      okResultWith(validSsn.value)
    } catch (e: RuntimeException) {
      when (e) {
        is IllegalArgumentException,
        is IllegalStateException,
        is DateTimeParseException -> invalidResultWith(ssn)
        else -> throw e
      }
    }
}
